### PR TITLE
[clangd] Support outgoing calls in call hierarchy

### PR DESCRIPTION
**Pull request by HighCommander4**

### CODE REVIEW

**Comment by HighCommander4** [comment]
(Not sure what happened here, the PR was somehow created while I was still writing the description. Anyways, updating the title now.)

**Comment by HighCommander4** [comment]
Since Phabricator has been [taken down](https://discourse.llvm.org/t/llvm-phabricator-turndown/76137), I'm resubmitting the patch implementing outgoing calls in call hierarchy that was previously posted at https://reviews.llvm.org/D93829.

Previous discussion can still be seen at the Phabricator link which now links to a static snapshot of the review.

**Comment by HighCommander4** [comment]
@sam-mccall review ping :)

I would particularly appreciate feedback on whether I should plan to set aside some time to implement the "deep lookup optimization" (from [this comment](https://reviews.llvm.org/D93829#4258101)), or whether the 2.5% increase in index memory usage (which I've achieved by implementing everything _except_ the deep lookup optimization from that comment) is acceptable.

**Comment by pidgeon777** [comment]
@sam-mccall also pinging to assist @HighCommander4 🙂

**Comment by HighCommander4** [comment]
Shopping around for some potential alternative reviewers :)

**Comment by pidgeon777** [comment]
I was thinking that it's a real pity this branch is so hard to get merged 😅 I think other than the review it would be 100% done? I mean, no big further modifications would be involved?

**Comment by HighCommander4** [comment]
> I think other than the review it would be 100% done? I mean, no big further modifications would be involved?

In terms of functionality, I'd say the patch is complete.

In terms of performance (and specifically memory usage), it's a matter of judgment of what we consider good enough. The current patch increases the memory usage of the index by 2.5% on the example workload of the LLVM codebase. That's not too bad (and a definite improvement over the 8.2% increase for the original version of the patch), but there is a possibility of doing a more involved rework of the index's in-memory representation (discussed under the name "deep lookup optimization" in the Phabricator thread) which would actually result in a net decrease in memory usage even with the feature.

One of the key pieces of feedback I'm hoping to get on this patch is whether the 2.5% memory usage increase is acceptable, or if we should consider the deep lookup optimization as a prerequisite.

**Comment by HighCommander4** [comment]
@kadircet perhaps you might be able to pick up this review?

Or, barring of a full review, your opinion on the directional question in [this comment](https://reviews.llvm.org/D93829#4654786) would be appreciated as well:

> how would you feel about proceeding with the patch in its current state, with the memory usage increase brought down from 8.2% to 2.5% thanks to the combination of the simple lookup optimization + RefKind filtering, and leaving the "deep lookup optimization" to be explored in a future change?

**Comment by pidgeon777** [comment]
Hopefully we're closer and closer to a merge, this really is a missing feature. I think this is the only LSP provider supporting just only one of the callHierarchy methods.

**Comment by HighCommander4** [comment]
@ckandeler, perhaps you would be willing to review this patch? I don't seem to have gotten any traction from my usual reviewers over the course of the past year.

File: clang-tools-extra/clangd/test/type-hierarchy.test
Comment (ckandeler) [inline clang-tools-extra/clangd/test/type-hierarchy.test:91]: What was the reasoning for removing this entirely?

File: clang-tools-extra/clangd/XRefs.cpp
Comment (ckandeler) [inline clang-tools-extra/clangd/XRefs.cpp:2349]: Can we get rid of the double negation? It just seems to obfuscate the logic for no particular reason.

File: clang-tools-extra/clangd/XRefs.cpp
Comment (ckandeler) [inline clang-tools-extra/clangd/XRefs.cpp:2342]: This comment doesn't really apply anymore.

File: clang-tools-extra/clangd/XRefs.cpp
Comment (i-garrison) [inline clang-tools-extra/clangd/XRefs.cpp:1755]: adding signature looks a bit more helpful, maybe one day add return type info too
```suggestion
  HI.detail = (S.Scope + S.Name + S.Signature).str();
```

File: clang-tools-extra/clangd/XRefs.cpp
Comment (HighCommander4) [reply clang-tools-extra/clangd/XRefs.cpp]: Reworded

File: clang-tools-extra/clangd/test/type-hierarchy.test
Comment (HighCommander4) [reply clang-tools-extra/clangd/test/type-hierarchy.test]: Huh; I don't recall. The test case seems to still be valid, so I restored it.

File: clang-tools-extra/clangd/XRefs.cpp
Comment (HighCommander4) [reply clang-tools-extra/clangd/XRefs.cpp:1756]: I agree, showing the signature here would be quite useful.

I actually started implementing this for incoming calls in https://github.com/clangd/clangd/issues/915, but ran into a complication (our current indexing process doesn't actually store the signature for all functions).

This is solvable with some indexer changes, but I'll defer to a fix for that issue, which will now (once this patch is merged) benefit both incoming and outgoing calls.

File: clang-tools-extra/clangd/XRefs.cpp
Comment (HighCommander4) [reply clang-tools-extra/clangd/XRefs.cpp:1704]: The difference between `declToHierarchyItem` and `symbolToHierarchyItem` is the former is used for symbols from the AST, and the latter for symbols from the index. They should both ideally produce the same result, but the implementation (and sometimes the amount of detail available in the respective inputs) is different.

**Comment by HighCommander4** [comment]
Thanks @ckandeler for the review! I addressed the comments in https://github.com/llvm/llvm-project/pull/77556/commits/bad4f7264433417763ea35bc1104fb087c8725e5.

File: clang-tools-extra/clangd/XRefs.cpp
Comment (i-garrison) [reply clang-tools-extra/clangd/XRefs.cpp:1704]: I see. Ideally there should be some recommended use model for `c++` which call hierarchy clients/visualisers (e.g. lsp4e) could implement to show expected function signatures. Like "obtain details field from call hierarchy item and use it as function signature" and then lsp4e could append details to name. We probably can later tune xrefs.cpp implementation to provide required details where needed. Note that function return value is still not available, and as far as I can see there is no place in protocol to put it in.

**Review by ckandeler — APPROVED** [review]
Appears to work fine.
Regarding the slight memory increase: If anyone feels strongly about it, they had plenty of time to chime in.

**Comment by pidgeon777** [comment]
First of all, I want to express my gratitude to all contributors working on the integration of changes and tests.

In today's development landscape, especially when working with AI-assisted coding, providing the best possible code context is crucial. The integration of outgoing calls functionality can significantly improve this aspect.

The main benefit of implementing outgoing calls would be context extension. This feature would:

- Allow developers to see function call relationships
- Help identify potential impacts when modifying functions
- Provide better context for AI-assisted development
- Improve code maintenance and refactoring workflows

When modifying a function, developers would have immediate visibility of all called functions that might need corresponding updates, reducing the risk of introducing bugs.

**Comment by HighCommander4** [comment]
> Appears to work fine. Regarding the slight memory increase: If anyone feels strongly about it, they had plenty of time to chime in.

Thanks! I'll go ahead and merge.

**Comment by HighCommander4** [comment]
Reverted inhttps://github.com/llvm/llvm-project/pull/117668 due to build breakage.

**Comment by HighCommander4** [comment]
> Reverted in #117668 due to build breakage.

The buildkite runs in this PR were green, but the PR was based on an older baseline, and some of the affected files have since changed in ways that require an update to the patch.

**Comment by HighCommander4** [comment]
Resubmitted with the build failures fixed at https://github.com/llvm/llvm-project/pull/117673.

**Pull request merged**
